src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/Models/ActiveDirectoryClient.cs

```diff
@@ -0,0 +1,818 @@
+                {
+                    options.SearchString = options.SearchString.TrimEnd('*');
+                    odataQuery = new Rest.Azure.OData.ODataQuery<ServicePrincipal>(s => s.DisplayName.StartsWith(options.SearchString));
+                }
+                else
+                {
+                    odataQuery = new Rest.Azure.OData.ODataQuery<ServicePrincipal>(s => s.DisplayName == options.SearchString);
+                }
+
+                return FilterServicePrincipals(odataQuery, first, skip);
+            }
+
+            return servicePrincipals;
+        }
+
+        public PSADUser CreateUser(UserCreateParameters userCreateParam)
+        {
+            return GraphClient.Users.Create(userCreateParam).ToPSADUser();
+        }
+
+        public PSADUser UpdateUser(string upnOrObjectId, UserUpdateParameters userUpdateParam)
+        {
+            GraphClient.Users.Update(upnOrObjectId, userUpdateParam);
+            return GraphClient.Users.Get(upnOrObjectId).ToPSADUser();
+        }
+
+        public void RemoveUser(string upnOrObjectId)
+        {
+            GraphClient.Users.Delete(upnOrObjectId);
+        }
+
+        public IEnumerable<PSADUser> FilterUsers(ADObjectFilterOptions options, ulong first = ulong.MaxValue, ulong skip = 0)
+        {
+            if (!string.IsNullOrEmpty(options.Id))
+            {
+                User user = null;
+                try
+                {
+                    user = GraphClient.Users.Get(Normalize(options.Id));
+                }
+                catch {  /* The user does not exist, ignore the exception. */ }
+
+                if (user != null)
+                {
+                    return new List<PSADUser> { user.ToPSADUser() };
+                }
+            }
+            else if (!string.IsNullOrEmpty(options.UPN) || !string.IsNullOrEmpty(options.Mail))
+            {
+                IPage<User> result = null;
+                try
+                {
+                    string upnOrMail = Normalize(options.UPN) ?? Normalize(options.Mail);
+                    var odataQuery = new Rest.Azure.OData.ODataQuery<User>(u => u.UserPrincipalName == upnOrMail);
+                    result = GraphClient.Users.List(odataQuery);
+                }
+                catch {  /* The user does not exist, ignore the exception. */ }
+
+                if (result != null)
+                {
+                    return result.Select(u => u.ToPSADUser());
+                }
+            }
+            else
+            {
+                Rest.Azure.OData.ODataQuery<User> odataQuery = null;
+                if (!string.IsNullOrEmpty(options.SearchString) && options.SearchString.EndsWith("*"))
+                {
+                    options.SearchString = options.SearchString.TrimEnd('*');
+                    odataQuery = new Rest.Azure.OData.ODataQuery<User>(u => u.DisplayName.StartsWith(options.SearchString));
```

Comment:
same comment.  ALso, it would be nice if we only hand to do this logic one time, rather than once per entity

Reply:
Same comment. I'm not sure if we can make this logic generic since the underlying types (`User`, `ServicePrincipal`, etc.) all derive from the same base type, but that base type doesn't have the `DisplayName` property that we're filtering over.